### PR TITLE
Suggested clarification and extensions

### PR DESCRIPTION
**Pull request by Goosth**
Using the editing to include suggested additions and clarifications on specific sections of the specification, and to create discussion.
(as recommended by @ianbjacobs).

Initially these are just comments and questions to test the view of the group. I will add suitable text once the concepts or extra examples are agreed to.

***

### :boom: Error: 400 Bad Request :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Aug 16, 2021, 11:30 AM UTC)_.

<details>
<summary>More</summary>

PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [CSS Spec Preprocessor](https://api.csswg.org/bikeshed/) - CSS Spec Preprocessor is the web service used to build Bikeshed specs.

:link: [Related URL](https://api.csswg.org/bikeshed/?url=https%3A%2F%2Fraw.githubusercontent.com%2Fw3c%2Fsecure-payment-confirmation%2F8946c61b3a55fd5f4d1858b057500dc58b081cb3%2Fspec.bs&md-warning=not%20ready)

```
Error running preprocessor, returned code: 2.
FATAL ERROR: Line 302 isn't indented enough (needs 1 indent) to be valid Markdown:
"  and both are legitimate?"
 ✘  Did not generate, due to fatal errors
```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20w3c/secure-payment-confirmation%23103.)._
</details>

### CODE REVIEW

File: spec.bs [hunk 1]
```diff
@@ -189,6 +189,12 @@ example flows and do not limit the scope of how the API can be used.
 This is the first-time flow, in which a new credential is created and stored by
 an issuing bank.
 
+**TODO**: Suggest we add an extra example, which might be more widely used: When a user navigates to bank.com for internet banking and 1) their account access
```
Comment (stephenmcgruer) [inline spec.bs:192]: I'm not opposed to such an example, though from an ideological view I continue to think we shouldn't let historical baggage of complex activations drag down WebAuthn/SPC credential creation :).

There is a slight complication: currently we are unsure whether the WebAuthn WG will allow an SPC credential to be entirely equivalent to a WebAuthn credential (i.e. make them interchangeable), or if they will require an SPC credential to have a special bit so that the RP knows this credential can be used in a third-party auth ceremony.

* If a special bit is required, this example will make a lot of sense. 
* If a special bit is **not** required, this example will just describe normal WebAuthn credential creation and thus be rather redundant imo.

File: spec.bs [hunk 2]
```diff
@@ -212,6 +218,10 @@ an issuing bank.
     credential is created and returned to the issuing bank who stores
     it in their server-side database associated with the user and
     payment instrument(s).
+    
+**TODO**: Define an Agent specific display for Enrollment, indicating that that Payment transactions will be supported (or that this is exclusively for payments)?
```
Comment (stephenmcgruer) [inline spec.bs:222]: Assuming 'Agent specific display' here means 'a browser UX', we (Chrome) oppose this currently. We are happy that the WebAuthn UX that would be shown is sufficient understanding for the user, and we believe that the RP is best suited to build and show any more specific UX that they wish to show regarding payments.

File: spec.bs [hunk 3]
```diff
@@ -212,6 +218,10 @@ an issuing bank.
     credential is created and returned to the issuing bank who stores
     it in their server-side database associated with the user and
     payment instrument(s).
+    
+**TODO**: Define an Agent specific display for Enrollment, indicating that that Payment transactions will be supported (or that this is exclusively for payments)? 
+
+**TODO**: Enhancement to consider: Allow the the user able to select if Payment should be enabled or not (the payment extension) or both?
```
Comment (stephenmcgruer) [inline spec.bs:224]: Again, we (Chrome) currently oppose this.

File: spec.bs [hunk 4]
```diff
@@ -229,6 +239,8 @@ const publicKey = {
   rp: {
     name: "Fancy Bank",
   },
+  
+  Q: Should we perhaps require that  rpId (domain) must be specified if this is inside an iFrame? For top level it's probably OK to use the top level domain?
```
Comment (stephenmcgruer) [inline spec.bs:243]: I'm not immediately sure why one would need to specify rpId here. I believe the WebAuthn spec says to use the callers (iframe's) domain:

https://w3c.github.io/webauthn/#dom-publickeycredentialcreationoptions-rp
```
Its value’s id member specifies the RP ID the credential should be scoped to. If omitted, its value
will be the CredentialsContainer object’s relevant settings object's origin's effective domain. See
§ 5.4.2 Relying Party Parameters for Credential Generation (dictionary PublicKeyCredentialRpEntity)
for further details.
```

Have I overlooked something?

File: spec.bs [hunk 5]
```diff
@@ -255,9 +267,12 @@ const publicKey = {
   authenticatorSelection: {
     userVerification: "required"
   },
+  
+  **TODO**: Extend definition (perhaps in future) to make this optional (so to only have userPresence: "required", and thereby support regions outside of europe) and remove friction of having to type in User PIN's?
```
Comment (stephenmcgruer) [inline spec.bs:271]: I think (though I'd need to double-check) that the spec doesn't currently require any particular `authenticatorSelection` options, though the user agent might (and Chrome currently does).

File: spec.bs [hunk 6]
```diff
@@ -255,9 +267,12 @@ const publicKey = {
...
 
-  timeout: 360000,  // 6 minutes
-
+  timeout: 60000,  // 1 minute
```
Comment (stephenmcgruer) [inline spec.bs:273]: No objection to the change to an example, I just stole this from WebAuthn examples anyway :P

File: spec.bs [hunk 7]
```diff
@@ -282,6 +297,16 @@ navigator.credentials.create({ publicKey })
 
 </div> <!-- non-normative -->
 
+**TODO**: Sugest we provide clarifying text/expand around the following 
+* Does the extension imply that this credential my ONLY be used for Payment, or does this mean the WebAuthn credential could be used for Login or Payment,
```
Comment (stephenmcgruer) [inline spec.bs:301]: Seems reasonable to clarify briefly. It can be used for both login and payment, if the RP wants to. They can also trivially not use it for login; if doing that we recommend they utilize a subdomain (e.g. `payment.bank.com`) to avoid polluting their credential space with credentials they wouldn't accept for login.

File: spec.bs [hunk 8]
```diff
@@ -282,6 +297,16 @@ navigator.credentials.create({ publicKey })
 
 </div> <!-- non-normative -->
 
+**TODO**: Sugest we provide clarifying text/expand around the following 
+* Does the extension imply that this credential my ONLY be used for Payment, or does this mean the WebAuthn credential could be used for Login or Payment, 
+  and both are legitimate?
+* Can the payment extension be 'added' once the webAuthn is created? Or must it be set at registration time? and can it be disabled too?
```
Comment (stephenmcgruer) [inline spec.bs:303]: Currently no, but I think we should think about this. As noted, a lot here depends on the WebAuthn WG's preference/requirements on whether SPC credential == WebAuthn credential, or if a separate bit (of some sort) must be set.

I *think* in general that WebAuthn doesn't allow modifying credentials (maybe I am incorrect), and so its generally a replace operation for something like an 'upgrade'.

File: spec.bs [hunk 9]
```diff
@@ -338,7 +363,9 @@ const request = new PaymentRequest([{
       icon: "https://fancybank.com/card-art.png",
     }, 
 
-    timeout: 360000,  // 6 minutes
+Q: Sugest we define that displayName is Mandatory and icon is optional?
```
Comment (stephenmcgruer) [inline spec.bs:366]: To be clear, you're suggesting a change to https://w3c.github.io/secure-payment-confirmation/#sctn-paymentcredentialinstrument-dictionary ?

I'm not sure offhand whether icon should be optional. Currently all users of SPC wish to pass an icon, and it's much easier to go required --> optional than the other way around, so I would by default lean to keeping both required.

File: spec.bs [hunk 10]
```diff
@@ -358,6 +387,8 @@ try {
 
   // response.data is a PublicKeyCredential, with a clientDataJSON that
   // contains the transaction data for verification by the issuing bank.
+  
+  Q: We need to define the format of Response.data somewhere in this doc, including what fields are present if the user accepts, of if they cancel/timeout.
```
Comment (stephenmcgruer) [inline spec.bs:391]: https://w3c.github.io/secure-payment-confirmation/#sctn-steps-to-respond-to-a-payment-request

File: spec.bs [hunk 11]
```diff
@@ -391,6 +422,8 @@ below and in [[#index-defined-elsewhere]].
     away from its Payment Request origins. For developers, this should
     improve feature detection, invocation, and other aspects of the
     API.</div>
+    
+    Q: Is the suggestion to create SPC specific methods, or roll everything into WebAuthn, with no custom-UI/logic in the browser agent?
```
Comment (stephenmcgruer) [inline spec.bs:426]: I'll leave the question of API shape for now, but note that we (Chrome) currently expect to always present payment-specific UX for an SPC authentication, no matter the API shape to call into it.

My personal understanding is that this NOTE only seeks to change the JavaScript code used to call SPC but none of the current behavior or UX would change.

File: spec.bs [hunk 12]
```diff
@@ -464,18 +499,27 @@ members:
...
     :: The number of milliseconds before the request to sign the transaction details times out. At most 1 hour.
+    
+    **TODO**: We need to define how this timeout is calculated. E.g. If the timeout is 60 seconds, and the uer spends 25 seconds on the SPC payment page in the browser, does the WebAuthn request then get initialized with 35 seconds or with 60 seconds? 
+    
+    **TODO**: What do we do if the SPC page is open for more than 60 seconds? Do we cancel without attempting WebAuthn?
 
     :  <dfn>payeeOrigin</dfn> member
     :: The origin of the payee that this SPC call is for (e.g., the merchant).
+    
+    **TODO**: Will this be validated in any way?
```
Comment (stephenmcgruer) [inline spec.bs:521]: It is impossible for the UA to validate this in the redirect case. This is identical to displayName and icon, where we are just relaying information from the SPC caller to the user, and then signing it to tell the RP that this is what we showed to the user. It's up to the RP to decide if it trusts the SPC caller to be doing a transaction for the payeeOrigin.

File: spec.bs [hunk 13]
```diff
@@ -464,18 +499,27 @@ members:
...
+    
+    **TODO**: We need to define how this timeout is calculated. E.g. If the timeout is 60 seconds, and the uer spends 25 seconds on the SPC payment page in the browser, does the WebAuthn request then get initialized with 35 seconds or with 60 seconds? 
+    
+    **TODO**: What do we do if the SPC page is open for more than 60 seconds? Do we cancel without attempting WebAuthn?
 
     :  <dfn>payeeOrigin</dfn> member
     :: The origin of the payee that this SPC call is for (e.g., the merchant).
+    
+    **TODO**: Will this be validated in any way? 
+    **TODO**: Will this be inserted by the browser agent if not provided?
```
Comment (stephenmcgruer) [inline spec.bs:522]: The conclusion in https://github.com/w3c/secure-payment-confirmation/pull/89 was to make it required, though I see you had a comment in that thread that I may have misunderstood (I thought you were saying the `topOrigin` field from your first bullet-point should be auto-filled, but perhaps you meant `payeeOrigin` should be auto-filled)

File: spec.bs [hunk 14]
```diff
@@ -533,6 +577,9 @@ The [=steps to respond to a payment request=] for this payment method, for a giv
     "{{AuthenticationExtensionsPaymentInputs/instrument}}" → |data|.{{SecurePaymentConfirmationRequest/instrument}}]».
 
     <div class="note">**TODO**: We do not have the RP origin at this step; maybe that should just go in the extension processing steps?</div>
+    Note: This is set in WebAuthn Authenticator, as part of creation? How do we validate/ensure we're not jumping to an incorrect domain? is this currently validated by the browser agent or by the CTAP protocol? Then we need to set it in some way?
+
+    Q: is payeeOrigin optional? What if it's not set? Do we allow it to be empty?
```
Comment (stephenmcgruer) [inline spec.bs:582]: It is not optional; https://w3c.github.io/secure-payment-confirmation/#dictdef-securepaymentconfirmationrequest

File: spec.bs [hunk 15]
```diff
@@ -533,6 +577,9 @@ The [=steps to respond to a payment request=] for this payment method, for a giv
     "{{AuthenticationExtensionsPaymentInputs/instrument}}" → |data|.{{SecurePaymentConfirmationRequest/instrument}}]».
 
     <div class="note">**TODO**: We do not have the RP origin at this step; maybe that should just go in the extension processing steps?</div>
+    Note: This is set in WebAuthn Authenticator, as part of creation? How do we validate/ensure we're not jumping to an incorrect domain? is this currently validated by the browser agent or by the CTAP protocol? Then we need to set it in some way?
```
Comment (stephenmcgruer) [inline spec.bs:580]: Can you rephrase this? I'm not sure what you're trying to ask here.

File: spec.bs [hunk 16]
```diff
@@ -544,6 +591,11 @@ The [=steps to respond to a payment request=] for this payment method, for a giv
     "{{PublicKeyCredentialRequestOptions/userVerification}}" → {{UserVerificationRequirement/required}},
     "{{PublicKeyCredentialRequestOptions/extensions}}" → |extensions|]».
 
+  **TODO**: Should the browser agent not extend the 'challenge' object passed in by putting it inside a new JSON structure whihc includes the payment amount and payee details?
```
Comment (stephenmcgruer) [inline spec.bs:594]: No, that is the purpose of the extension to the CollectedClientData, to explicitly avoid hacking the `challenge`: https://w3c.github.io/secure-payment-confirmation/#sctn-collectedclientpaymentdata-dictionary

File: spec.bs [hunk 17]
```diff
@@ -544,6 +591,11 @@ The [=steps to respond to a payment request=] for this payment method, for a giv
     "{{PublicKeyCredentialRequestOptions/userVerification}}" → {{UserVerificationRequirement/required}},
     "{{PublicKeyCredentialRequestOptions/extensions}}" → |extensions|]».
 
+  **TODO**: Should the browser agent not extend the 'challenge' object passed in by putting it inside a new JSON structure whihc includes the payment amount and payee details? 
+  
+  **TODO**: Userverification + " required' would probably also come from the WebAuthn value. We cannot have a different auth being
```
Comment (stephenmcgruer) [inline spec.bs:596]: Cut-off sentence?

File: spec.bs [hunk 18]
```diff
@@ -559,6 +611,8 @@ The [=steps to respond to a payment request=] for this payment method, for a giv
     → |publicKeyOpts|]».
 
     Note: This triggers [[webauthn-3]]'s [[webauthn-3#sctn-getAssertion|Get]] behavior
+    
+    Q: Which version of WebAuthn will SPC require? Can we make it work for Level 1, or is Level 2 or 3 required?
```
Comment (stephenmcgruer) [inline spec.bs:615]: I think L2 might be required for discoverable credentials?

File: spec.bs [hunk 19]
```diff
@@ -575,6 +629,8 @@ this extension allows a third-party to perform an authentication ceremony on
 behalf of the [=Relying Party=], and also adds transaction information to the
 signed cryptogram.
 
+Q: Where is the adding of Transaction information defined? I think we should define it in more detail in this spec?
```
Comment (stephenmcgruer) [inline spec.bs:632]: https://w3c.github.io/secure-payment-confirmation/#sctn-payment-extension-enrollment

Under 'Client extension processing (authentication)', where it modifies step 9.

File: spec.bs [hunk 20]
```diff
@@ -586,6 +642,8 @@ directly; for authentication the extension can only be accessed via
 :  Operation applicability
 :: [=registration extension|Registration=] and [=authentication extension|authentication=]
 
+Q: Why does this apply to Registration? Should it not just be for authentication?
```
Comment (stephenmcgruer) [inline spec.bs:645]: As per 'Client extension processing (registration)' below, it needs to relax the ban on creation in a cross-origin iframe.

File: spec.bs [hunk 21]
```diff
@@ -595,11 +653,14 @@ directly; for authentication the extension can only be accessed via
     dictionary AuthenticationExtensionsPaymentInputs {
       boolean isPayment;
 
+Q: When would this be set to False? Is this not true by the virtue of being present?
```
Comment (stephenmcgruer) [inline spec.bs:656]: Yes, this is just an awkward outcome of WebIDL treating an empty dict as false (afaik). So if during registration you did:

```
const publicKeyCreationOptions = {
  ...
  extensions: {
    payment: {},
  },
};
navigator.credentials.create({publicKey: publicKeyCreationOptions});
```

Then afaik the payment extension is actually *not* set. So you need some way to indicate 'no I do want this'.

As per the note under https://w3c.github.io/secure-payment-confirmation/#sctn-enrollment, if WebAuthn WG allow credential creation in a 3p context, and if we get to a point where the browser doesn't have to cache the credentials at creation time, this extension will no longer need to be registration time too, and this field can be dropped.

(`isPayment` definitely needs  better documentation. It is mentioned here https://w3c.github.io/secure-payment-confirmation/#dom-authenticationextensionspaymentinputs-ispayment but not well)

File: spec.bs [hunk 22]
```diff
@@ -642,6 +703,8 @@ directly; for authentication the extension can only be accessed via
          (see [[#sctn-permissions-policy]]). We could additionally require and
          consume a [=transient activation=] here, if we felt the permission policy
          is not sufficient.
+         
+         Note: I feel we must still check that the ID's corresponds to the domain that was provided; you cannot use a key with Bank2 if Bank1 claims to be asking for it? The RP must be included in the signature at least.
```
Comment (stephenmcgruer) [inline spec.bs:707]: The RP is in the signature; https://w3c.github.io/secure-payment-confirmation/#dom-collectedclientadditionalpaymentdata-rp

File: spec.bs [hunk 23]
```diff
@@ -656,6 +719,8 @@ directly; for authentication the extension can only be accessed via
         1. Skip step 6.1, which compares *options.rpId* to *effectiveDomain*
 
             Note: This enables cross-domain authentication ceremonies; see [[#sctn-use-case-merchant-authentication]].
+            
+            Q: Should we not at least compare to the claimed Bank Identity? How else will we know the value displayed is actually linked to the credential?
```
Comment (stephenmcgruer) [inline spec.bs:723]: Because currently the caller of SPC doesn't set the RP origin shown to the user. There is no input to https://w3c.github.io/secure-payment-confirmation/#sctn-securepaymentconfirmationrequest-dictionary where one can 'claim' the RP value.

How the user agent actually *gets* the RP's identity from the credential is an outstanding issue for the spec. In the Chrome impl, we cache it at creation time. In the future, I hope to retrieve it via the Conditional UI concept in WebAuthn (which will also require the ability to fetch an RP for a given credential).

We *could* ask the caller to make an assertion about what domain it is, but I think given that there are technical solutions to avoid that, we should. If we do ultimately ask the caller to make such a claim, then yes we should verify that once we know the value.

File: spec.bs [hunk 24]
```diff
@@ -724,6 +791,8 @@ fields:
     :: The instrument information that was displayed to the user.
 </dl>
 
+Q: We should probably state that these fields must be displayed to the user.
```
Comment (stephenmcgruer) [inline spec.bs:794]: See the big TODO in https://w3c.github.io/secure-payment-confirmation/#sctn-transaction-confirmation-ux ;)

File: spec.bs (hunk 1 above)
Comment (Goosth) [reply spec.bs:192]: If we're aligning to Webauthn, then I would suggest we remove the detailed explanation and flows. It would be better to rather just refer to the WebAuthn document and process for Enrollment.

We would then just
* give high level explanation on the use-cases (the two mentioned)
* Re-iterate that this is standard WebAuthn and there is no cross domain capability in registration, so the standard WebAuthn process will be followed with no UI.

File: spec.bs (hunk 2 above)
Comment (Goosth) [reply spec.bs:222]: Ok. Let's then just state that (see my previous comment).

File: spec.bs (hunk 4 above)
Comment (Goosth) [reply spec.bs:243]: Given that this is a Registration and there is no cross domain and no custom display, I agree that this should be fine.

File: spec.bs (hunk 8 above)
Comment (Goosth) [reply spec.bs:303]: That's my understanding as well. Let's state this is the situation right now. (Do you think we need to verify this with WebAuthn team?)

File: spec.bs (hunk 12 above)
Comment (Goosth) [reply spec.bs:521]: Thanks.

File: spec.bs (hunk 5 above)
Comment (Goosth) [reply spec.bs:271]: So would Chrome then be deemed as compliant if it only supports a subset?
Is it your intention to support the other options as well in future; just not initially?

File: spec.bs (hunk 13 above)
Comment (Goosth) [reply spec.bs:522]: I guess both should be filled. But agree on the comment that it's not provable. Perhaps better to, similar to icon and displayName and make it mandatory.

File: spec.bs (hunk 24 above)
Comment (Goosth) [reply spec.bs:794]: Saw it. :-)

File: spec.bs (hunk 18 above)
Comment (Goosth) [reply spec.bs:615]: Will you be able tp make it work without that or is discoverable credentials a minimum requirement?

File: spec.bs (hunk 1 above)
Comment (ianbjacobs) [reply spec.bs:192]: Hi @goosth, I think there is a cross-origin iframe registration capability in SPC, without special UI in the current Chrome implementation.

File: spec.bs (hunk 8 above)
Comment (stephenmcgruer) [reply spec.bs:303]: I will be speaking at the WebAuthn WG tomorrow afternoon EST, and I expect a resolution to come out of that.

File: spec.bs (hunk 5 above)
Comment (stephenmcgruer) [reply spec.bs:271]: > So would Chrome then be deemed as compliant if it only supports a subset?

Nope! :D

> Is it your intention to support the other options as well in future; just not initially?

I think this will go one of two ways: either we will hear no demand for other options and may restrict SPC for simplicity and/or clarity, or we will hear a compelling use-case and Chrome will add support for other options.

File: spec.bs (hunk 18 above)
Comment (stephenmcgruer) [reply spec.bs:615]: I would consider them a minimum requirement, in order to move from a browser-based cache to a platform-based cache and thus avoid SPC credentials being per-browser on a given device. See https://github.com/w3c/secure-payment-confirmation/issues/92

File: spec.bs (hunk 1 above)
Comment (Goosth) [reply spec.bs:192]: Agreed. But that is a WebAuthn standard. You still need the iframe to be opened in the RP's domain and the iframe to be given the correct permissions by the merchant. 
It's not using the SPC superpower of being called by the merchant and creating a key in the RP's domain.

File: spec.bs (hunk 5 above)
Comment (Goosth) [reply spec.bs:271]: I guess in that case then we'll need to pair back the spec at a later stage
@ianbjacobs : What is the best advice 
1.  have the spec adhere more to Chrome's initial implementation and note this as a future extension
2. have the spec be more generic and not have Chrome fully implement it?

File: spec.bs (hunk 22 above)
Comment (Goosth) [reply spec.bs:707]: If the claimed RP is different than the RP that created it, why would the Browser/WebAuthn not also verify it?
First check that the RP is correct before asking the Authenticator to sign it?

Comment (stephenmcgruer) [reply spec.bs:707]: Sorry, I should have quoted to indicate the part I was replying to: "The RP must be included in the signature at least."

The caller of SPC makes no claim on the RP (it's not [an input to the API](https://w3c.github.io/secure-payment-confirmation/#sctn-securepaymentconfirmationrequest-dictionary)). They pass in a list of credential IDs, which do not contain RP information. The browser/authenticator* then matches the credential IDs to the underlying RP, which is what is signed.

(One thing the spec doesn't deal with today which we should fix - what if you pass in credential IDs from multiple RPs, what happens. This is a silly edge-case that doesn't really matter, but we should spec it to just fail in a privacy-preserving way probably).

* Today via the browser-stored metadata from credential creation time, in the long-term this should be queryable from the authenticator.

**Comment by ianbjacobs** [comment]
@Goosth, I haven't checked in lately to see whether we've addressed issues here to your satisfaction.

**Comment by ianbjacobs** [comment]
Closing this issue due to inactivity and the fact that many points have been taken into account separately.
